### PR TITLE
Feat/macro physical particles

### PR DESCRIPTION
**Pull request by cremebrule**

### CODE REVIEW

File: omnigibson/systems/macro_particle_system.py [hunk 1]
```diff
@@ -43,14 +29,13 @@ class MacroParticleSystem(BaseSystem):
...
+    _particle_object = None
 
     # dict, array of particle objects, mapped by their prim names
     particles = None
 
-    # Scaling factor to sample from when generating a new particle
-    min_scale = None              # (x,y,z) scaling
-    max_scale = None              # (x,y,z) scaling
+    # Counter to increment monotonically as we add more particles
+    _particle_counter = None
```
Comment (ChengshuLi) [inline omnigibson/systems/macro_particle_system.py:38]: If we want to manually keep track of `_particle_counter`, than we also need to add it to load/dump_state?

File: omnigibson/systems/macro_particle_system.py [hunk 2]
```diff
@@ -155,95 +140,60 @@ def particle_name_prefix(cls):
...
-            poses=[particle.get_local_pose() for particle in cls.particles.values()],
-            scales=[particle.scale for particle in cls.particles.values()],
-            template_pose=cls.particle_object.get_local_pose() if cls.particle_object is not None else None,
-            template_scale=cls.particle_object.scale if cls.particle_object is not None else None,
-        )
+        state = super()._dump_state()
+
+        state["scales"] = np.array([particle.scale for particle in cls.particles.values()])
+        state["template_pose"] = np.concatenate(cls._particle_object.get_local_pose())
+        state["template_scale"] = cls._particle_object.scale
```
Comment (ChengshuLi) [inline omnigibson/systems/macro_particle_system.py:153]: Sorry I know these are from before, but why do we need template_pose and template_scale again?

File: omnigibson/systems/macro_particle_system.py [hunk 3]
```diff
@@ -725,35 +469,18 @@ def generate_group_particles(
...
             )
 
             # Add to group
             cls._group_particles[group][particle.name] = particle
             cls._particles_info[particle.name] = dict(obj=cls._group_objects[group], link=link)
 
-            # Update particle local matrix
-            cls._particles_local_mat[particle.name] = cls._compute_particle_local_mat(name=particle.name)
+            # Set the pose
+            cls.set_particle_position_orientation(idx=-1, position=position, orientation=orientation, local=False)
```
Comment (ChengshuLi) [inline omnigibson/systems/macro_particle_system.py:480]: `idx=-1`?

File: omnigibson/systems/macro_particle_system.py [hunk 4]
```diff
@@ -812,92 +539,177 @@ def generate_group_particles_on_object(cls, group, max_samples, min_samples_for_
...
-            particle_local_poses_batch[i] = cls._particles_local_mat[name]
-
-        # Compute once
-        global_poses = np.matmul(link_tfs_batch, particle_local_poses_batch)
+                - (n, 3)-array: per-particle (x,y,z) position
+                - (n, 4)-array: per-particle (x,y,z,w) quaternion orientation
+        """
+        n_particles = len(particles)
+        if local:
+            global_poses = np.zeros((n_particles, 4, 4))
```
Comment (ChengshuLi) [inline omnigibson/systems/macro_particle_system.py:557]: `global_poses` is a bit misleading here? In general I think this mixture of global and local is extremely error-prone. Let's discuss in person to see if there is a better way.

File: omnigibson/systems/macro_particle_system.py [hunk 5]
```diff
@@ -812,92 +539,177 @@ def generate_group_particles_on_object(cls, group, max_samples, min_samples_for_
...
+            link_tfs_batch = np.zeros((n_particles, 4, 4))
+            particle_local_poses_batch = np.zeros_like(link_tfs_batch)
+            for i, name in enumerate(particles):
+                link = cls._particles_info[name]["link"]
+                if link in link_tfs:
+                    link_tf = link_tfs[link]
+                else:
+                    link_tf = T.pose2mat(link.get_position_orientation())
+                    link_tfs[link] = link_tf
+                link_tfs_batch[i] = link_tf
```
Comment (ChengshuLi) [inline omnigibson/systems/macro_particle_system.py:572]: nit: changed to
```
if link not in link_tfs:
    link_tfs[link] = T.pose2mat(link.get_position_orientation())

link_tfs_batch[i] = link_tfs[link]
```

File: omnigibson/systems/macro_particle_system.py [hunk 6]
```diff
@@ -812,92 +539,177 @@ def generate_group_particles_on_object(cls, group, max_samples, min_samples_for_
...
+    def get_group_particles_position_orientation(cls, group, local=False):
+        return cls._compute_batch_particles_position_orientation(particles=cls._group_particles[group], local=local)
+
+    @classmethod
+    def get_particle_position_orientation(cls, idx, local=False):
+        name = list(cls.particles.keys())[idx]
+        if local:
+            val = cls.particles[name].get_local_pose()
+        else:
+            # First, get local pose, scale it by the parent link's scale, and then convert into a matrix
```
Comment (ChengshuLi) [inline omnigibson/systems/macro_particle_system.py:595]: is this comment still accurate? how to handle scale here?

File: omnigibson/systems/macro_particle_system.py [hunk 7]
```diff
@@ -1129,7 +932,344 @@ def _deserialize(cls, state):
...
+        if not get_prim_at_path(prim_path):
+            omni.kit.commands.execute(
+                "CopyPrim",
+                path_from=cls._particle_object.prim_path,
+                path_to=prim_path,
+            )
+            # Apply RigidBodyAPI to it so it is subject to physics
+            prim = get_prim_at_path(prim_path)
+            UsdPhysics.RigidBodyAPI.Apply(prim)
+            # PhysxSchema.PhysxRigidBodyAPI.Apply(prim)
```
Comment (ChengshuLi) [inline omnigibson/systems/macro_particle_system.py:974]: Is this line needed?

File: omnigibson/systems/macro_particle_system.py [hunk 8]
```diff
@@ -1129,7 +932,344 @@ def _deserialize(cls, state):
...
+
+        # If sim is already playing, refresh particles immediately
+        if og.sim.is_playing():
+            cls._refresh_particles_view()
+
+    @classmethod
+    def _load_new_particle(cls, prim_path, name):
+        # We copy the template prim and generate the new object if the prim doesn't already exist, otherwise we
+        # reference the pre-existing one
+        if not get_prim_at_path(prim_path):
```
Comment (ChengshuLi) [inline omnigibson/systems/macro_particle_system.py:965]: When would we call `_load_new_particle` to a prim_path that already exists? Is it safer to raise an error?

File: omnigibson/systems/macro_particle_system.py [hunk 9]
```diff
@@ -1129,7 +932,344 @@ def _deserialize(cls, state):
...
+        return CollisionVisualGeomPrim(prim_path=prim_path, name=name)
+
+    @classmethod
+    def set_particle_template_object(cls, obj):
+        # Run super method
+        super().set_particle_template_object(obj=obj)
+
+        # Make sure min and max scale are identical
+        assert np.all(cls.min_scale == cls.max_scale), \
+            "Min and max scale should be identical for MacroPhysicalParticleSystem!"
```
Comment (ChengshuLi) [inline omnigibson/systems/macro_particle_system.py:984]: hmm interesting, would it be more realistic if we actually allow different scales?

File: omnigibson/systems/macro_particle_system.py [hunk 10]
```diff
@@ -1129,7 +932,344 @@ def _deserialize(cls, state):
...
+
+        # Compute particle radius
+        vertices = np.array(cls._particle_object.get_attribute("points")) * cls.max_scale.reshape(1, 3)
+        cls._particle_offset, cls._particle_radius = trimesh.nsphere.minimum_nsphere(trimesh.Trimesh(vertices=vertices))
```
Comment (ChengshuLi) [inline omnigibson/systems/macro_particle_system.py:988]: Does this need to be multipled by scale?

File: omnigibson/systems/macro_particle_system.py [hunk 11]
```diff
@@ -1129,7 +932,344 @@ def _deserialize(cls, state):
...
+            pos, ori = tfs[:, :3], tfs[:, 3:]
+            pos = pos + T.quat2mat(ori) @ cls._particle_offset
+        else:
+            pos, ori = np.array([]).reshape(0, 3), np.array([]).reshape(0, 4)
+        return pos, ori
+
+    @classmethod
+    def get_particle_position_orientation(cls, idx, local=False):
+        positions, orientations = cls.get_particles_position_orientation()
+        return (positions[idx], orientations[idx]) if cls.n_particles > 0 else (positions, orientations)
```
Comment (ChengshuLi) [inline omnigibson/systems/macro_particle_system.py:1038]: hmm is this the expected behavior? Maybe should raise an exception instead?

File: omnigibson/systems/macro_particle_system.py [hunk 12]
```diff
@@ -1129,7 +932,344 @@ def _deserialize(cls, state):
...
+        """
+        Grab particle @idx's global linear and angular velocities
+
+        Returns:
+            2-tuple:
+                - 3-array: particle (x, y, z) linear velocity in the world frame
+                - 3-array: particle (ax, ay, az) angular velocity in the world frame
+        """
+        lin_vel, ang_vel = cls.get_particles_velocities()
+        return (lin_vel[idx], ang_vel[idx]) if cls.n_particles > 0 else lin_vel, ang_vel
```
Comment (ChengshuLi) [inline omnigibson/systems/macro_particle_system.py:1084]: same

File: omnigibson/systems/macro_particle_system.py [hunk 13]
```diff
@@ -335,10 +258,47 @@ def remove_particle(cls, name):
...
+    ):
+        particle_names = tuple(cls.particles.keys())
+        for idx in idxs:
+            cls.remove_particle(particle_names[idx])
+
+    @classmethod
+    def generate_particles(
+            cls,
+            positions,
+            velocities=None,
```
Comment (ChengshuLi) [inline omnigibson/systems/macro_particle_system.py:275]: velocities are never used?

File: omnigibson/systems/macro_particle_system.py [hunk 14]
```diff
@@ -1129,7 +932,344 @@ def _deserialize(cls, state):
...
+        # keep the first (N - n_new) values
+        current_lin_vels, current_ang_vels = cls.get_particles_velocities()
+
+        # Update the tensors
+        n_particles = len(positions)
+        velocities = np.zeros((n_particles, 3)) if velocities is None else velocities
+        angular_velocities = np.zeros_like(velocities) if angular_velocities is None else angular_velocities
+
+        velocities = np.concatenate([current_lin_vels[:-n_particles], velocities], axis=0)
+        angular_velocities = np.concatenate([current_ang_vels[:-n_particles], angular_velocities], axis=0)
```
Comment (ChengshuLi) [inline omnigibson/systems/macro_particle_system.py:1155]: So at this point, I assume the newly added particles have zero velocities by default?

File: omnigibson/systems/macro_particle_system.py [hunk 15]
```diff
@@ -1163,6 +1303,25 @@ def _deserialize(cls, state):
 )
 
 
+MacroPhysicalParticleSystem.create(
+    name="raspberry",
+    particle_density=800.0,
+    create_particle_template=lambda prim_path, name: og.objects.DatasetObject(
+        prim_path=prim_path,
+        # usd_path=os.path.join(gm.ASSET_PATH, "models", "stain", "stain.usd"),
```
Comment (ChengshuLi) [inline omnigibson/systems/macro_particle_system.py:1311]: remove?

File: omnigibson/systems/macro_particle_system.py [hunk 16]
```diff
@@ -1163,6 +1303,25 @@ def _deserialize(cls, state):
...
+        name=name,
+        # class_id=SemanticClass.DIRT,
```
Comment (ChengshuLi) [inline omnigibson/systems/macro_particle_system.py:1313]: oh we should start handling segmentation mask for systems correctly..

File: omnigibson/systems/macro_particle_system.py [hunk 17]
```diff
@@ -1129,7 +932,344 @@ def _deserialize(cls, state):
...
+    def create(cls, name, create_particle_template, particle_density, scale, **kwargs):
+        """
+        Utility function to programmatically generate monolithic visual particle system classes.
+
+        Note: If using super() calls in any functions, we have to use slightly esoteric syntax in order to
+        accommodate this procedural method for using super calls
+        cf. https://stackoverflow.com/questions/22403897/what-does-it-mean-by-the-super-object-returned-is-unbound-in-python
+            Use: super(cls).__get__(cls).<METHOD_NAME>(<KWARGS>)
+
+        Note: Scale is automatically inferred from the particle object that is loaded!
```
Comment (ChengshuLi) [inline omnigibson/systems/macro_particle_system.py:1170]: Is this comment accurate? I thought scale is [1, 1, 1] if not specified.

File: omnigibson/systems/macro_particle_system.py [hunk 18]
```diff
@@ -1163,6 +1303,25 @@ def _deserialize(cls, state):
...
+        visual_only=True,
+        include_default_states=False,
+        category="raspberry",
+        model="spkers",
+    ),
+    scale=np.ones(3) * 5.0,
+)
+
+
 # GrassSystem = VisualParticleSystem.create(
```
Comment (ChengshuLi) [inline omnigibson/systems/macro_particle_system.py:1325]: Maybe remove the grass system for now?

File: omnigibson/systems/micro_particle_system.py [hunk 19]
```diff
@@ -725,21 +689,26 @@ def _create_particle_prototypes(cls):
         raise NotImplementedError()
 
     @classmethod
-    def check_in_contact(cls, positions):
+    def delete_particles(
+            cls,
+            idxs,
+            instancer_idn=None,
```
Comment (ChengshuLi) [inline omnigibson/systems/micro_particle_system.py:695]: are we getting rid of `instancer_idn`?

File: omnigibson/systems/micro_particle_system.py [hunk 20]
```diff
@@ -777,7 +744,7 @@ def generate_particles(
             self_collision (bool): Whether to enable particle-particle collision within the set
                 (as defined by @particle_group) or not
             prototype_indices (None or list of int): If specified, should specify which prototype should be used for
-                each particle. If None, will use all 0s (i.e.: the first prototype created)
+                each particle. If None, will randomly sample from all available prototypes
```
Comment (ChengshuLi) [inline omnigibson/systems/micro_particle_system.py:747]: hmm why this change?

File: omnigibson/systems/micro_particle_system.py [hunk 21]
```diff
@@ -1635,6 +1556,18 @@ class Cloth(MicroParticleSystem):
     """
     Particle system class to simulate cloth.
     """
+    @classproperty
+    def particle_radius(cls):
+        # Magic number from omni tutorials
+        # See https://docs.omniverse.nvidia.com/prod_extensions/prod_extensions/ext_physics.html#offset-autocomputation
+        # Also https://nvidia-omniverse.github.io/PhysX/physx/5.1.3/docs/ParticleSystem.html#particle-system-configuration
+        return 0.99 * cls.particle_contact_offset
```
Comment (ChengshuLi) [inline omnigibson/systems/micro_particle_system.py:1564]: Is this newly added or moved from somewhere else?

File: omnigibson/systems/micro_particle_system.py [hunk 22]
```diff
@@ -1680,6 +1613,7 @@ def clothify_mesh_prim(cls, mesh_prim, remesh=True, particle_distance=None):
                     # Terminate anyways, but don't fail
                     log.warn("Failed to sufficiently remesh cloth. "
                              "The generated cloth may not have evenly distributed particles.")
+                    break
```
Comment (ChengshuLi) [inline omnigibson/systems/micro_particle_system.py:1616]: Regarding `avg_edge_percentage_mismatch`, I wonder whether checking "avg_edge_length" is enough. Ultimately we want the edges to have similar length and that length is close to `particle_distance`. Maybe we should use that stop criterion instead?

**Review by ChengshuLi — COMMENTED** [review]
Awesome work!! Let's go over this together in person.

File: omnigibson/systems/system_base.py [hunk 23]
```diff
@@ -134,6 +373,536 @@ def __init__(self):
...
+        # Store local poses since particles are attached to moving bodies
+        return True
+
+    @classproperty
+    def scale_relative_to_parent(cls):
+        """
+        Returns:
+            bool: Whether or not particles should be scaled relative to the group's parent object
+        """
+        return False
```
[the remark this answers is not in the document]
Comment (ChengshuLi) [reply omnigibson/systems/system_base.py:423]: This comes from per-substance annotation from jsons.

File: omnigibson/systems/system_base.py [hunk 24]
```diff
@@ -119,9 +223,144 @@ def get_active_systems(cls):
         """
         return {system.name: system for system in SYSTEM_REGISTRY.objects if issubclass(system, cls)}
 
+    @classmethod
+    def set_scale_limits(cls, minimum=None, maximum=None):
```
[the remark this answers is not in the document]
Comment (ChengshuLi) [reply omnigibson/systems/system_base.py:227]: maybe putting some comments.

File: omnigibson/systems/system_base.py [hunk 25]
```diff
@@ -134,6 +373,536 @@ def __init__(self):
...
+            float: Contact radius for the particles to be generated, for the purpose of estimating contacts
+        """
+        raise NotImplementedError()
+
+    @classmethod
+    def check_in_contact(cls, positions):
+        """
+        Checks whether each particle specified by @particle_positions are in contact with any rigid body.
+
+        NOTE: This is a rough proxy for contact, given @positions. Should not be taken as ground truth.
```
[the remark this answers is not in the document]
Comment (ChengshuLi) [reply omnigibson/systems/system_base.py:756]: maybe try overlap_shape?

[the remark this answers is not in the document]
Comment (ChengshuLi) [reply omnigibson/systems/system_base.py:756]: explain why and when it's a proxy.

File: omnigibson/systems/macro_particle_system.py (hunk 2 above)
Comment (ChengshuLi) [reply omnigibson/systems/macro_particle_system.py:153]: delete, see Slack

File: omnigibson/systems/macro_particle_system.py (hunk 6 above)
Comment (ChengshuLi) [reply omnigibson/systems/macro_particle_system.py:595]: Mention that `_particles_local_mat` already takes scale into account.

File: omnigibson/systems/macro_particle_system.py (hunk 10 above)
Comment (ChengshuLi) [reply omnigibson/systems/macro_particle_system.py:988]: Add a comment in get/set_position_orientation saying it's the center of the sphere, rather than the particle frame.

File: omnigibson/systems/macro_particle_system.py (hunk 17 above)
Comment (ChengshuLi) [reply omnigibson/systems/macro_particle_system.py:1170]: Make sure always infer particle radius from (scaled) USD, ideally scale is [1, 1, 1] assuming the asset has the correct scale.

File: omnigibson/object_states/covered.py [hunk 26]
```diff
@@ -114,8 +113,7 @@ def _set_value(self, system, new_value):
                     )
                 else:
                     # We delete all particles touching this object
-                    for inst, particle_idxs in self.obj.states[ContactParticles].get_value(system).items():
-                        inst.remove_particles(idxs=list(particle_idxs))
+                    system.delete_particles(idxs=list(self.obj.states[ContactParticles].get_value(system)))
```
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/object_states/covered.py:116]: fixed

File: omnigibson/object_states/particle_modifier.py [hunk 27]
```diff
@@ -282,8 +282,9 @@ def overlap_callback(hit):
 
             # Make sure the mesh is translated so that its tip lies at the metalink origin, and rotated so the vector
             # from tip to tail faces the positive x axis
-            z_offset = self._projection_mesh_params["extents"][2] if self._projection_mesh_params["type"] == "Cone" \
-                else self._projection_mesh_params["extents"][2] / 2
+            # Both Cone and Cylinder have their frames located at the center of their shapes as of IsaacSim 2022.2.1,
+            # so we offset by half their height
+            z_offset = self._projection_mesh_params["extents"][2] / 2
```
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/object_states/particle_modifier.py:287]: done

File: omnigibson/object_states/contains.py [hunk 28]
```diff
@@ -56,23 +56,23 @@ def _get_value(self, system):
...
+                    unit_z[:, -1, :] = m.VISUAL_PARTICLE_OFFSET
+                    checked_positions = (T.quat2mat(quats) @ unit_z).reshape(-1, 3) + raw_positions
+            elif is_physical_particle_system(system_name=system.name):
+                raw_positions = system.get_particles_position_orientation(local=False)[0]
                 checked_positions = raw_positions
-        else:
-            raise ValueError(f"Invalid system {system} received for getting Covered state!"
-                             f"Currently, only VisualParticleSystems and PhysicalParticleSystems are supported.")
+            else:
+                raise ValueError(f"Invalid system {system} received for getting Covered state!"
```
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/object_states/contains.py:74]: fixed

File: omnigibson/systems/system_base.py [hunk 29]
```diff
@@ -119,9 +223,144 @@ def get_active_systems(cls):
...
+
+    @classmethod
+    def get_particles_position_orientation(cls, local=False):
+        """
+        Computes all particles' positions and orientations that belong to this system
+
+        Note: This is more optimized than doing a for loop with self.get_particle_position_orientation()
+
+        Args:
+            local (bool): Whether to get pose in the particle's local frame or not
```
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/systems/system_base.py:250]: done

File: omnigibson/systems/system_base.py [hunk 30]
```diff
@@ -134,6 +373,536 @@ def __init__(self):
...
+            bbox_upper_limit = np.clip(
+                bbox_upper_limit_from_aabb,
+                m.BBOX_UPPER_LIMIT_MIN,
+                m.BBOX_UPPER_LIMIT_MAX,
+            )
+
+            # Convert these into scaling factors for the x and y axes for our particle object
+            particle_bbox = cls.particle_object.aabb_extent
+            minimum = np.array([bbox_lower_limit / particle_bbox[0], bbox_lower_limit / particle_bbox[1], 1.0])
+            maximum = np.array([bbox_upper_limit / particle_bbox[0], bbox_upper_limit / particle_bbox[1], 1.0])
```
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/systems/system_base.py:585]: yes, but it doesn't make sense in many cases. z shouldn't get scaled so that "flat" substances become bulgy, like stains, right?

File: omnigibson/systems/system_base.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/systems/system_base.py:227]: fixed

File: omnigibson/systems/system_base.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/systems/system_base.py:756]: done

File: omnigibson/systems/macro_particle_system.py (hunk 1 above)
Comment (cremebrule) [reply omnigibson/systems/macro_particle_system.py:38]: done

File: omnigibson/systems/macro_particle_system.py (hunk 2 above)
Comment (cremebrule) [reply omnigibson/systems/macro_particle_system.py:153]: done

File: omnigibson/systems/macro_particle_system.py (hunk 4 above)
Comment (cremebrule) [reply omnigibson/systems/macro_particle_system.py:557]: fixed

File: omnigibson/systems/macro_particle_system.py (hunk 5 above)
Comment (cremebrule) [reply omnigibson/systems/macro_particle_system.py:572]: done

File: omnigibson/systems/macro_particle_system.py (hunk 6 above)
Comment (cremebrule) [reply omnigibson/systems/macro_particle_system.py:595]: done

File: omnigibson/systems/macro_particle_system.py (hunk 7 above)
Comment (cremebrule) [reply omnigibson/systems/macro_particle_system.py:974]: no, removed

File: omnigibson/systems/macro_particle_system.py (hunk 8 above)
Comment (cremebrule) [reply omnigibson/systems/macro_particle_system.py:965]: actually, it might already exist. this could happen during the `sync_particle_groups` call, where there may already be the correct particles on the stage, and we want to simply move them around

File: omnigibson/systems/macro_particle_system.py (hunk 10 above)
Comment (cremebrule) [reply omnigibson/systems/macro_particle_system.py:988]: done

File: omnigibson/systems/macro_particle_system.py (hunk 11 above)
Comment (cremebrule) [reply omnigibson/systems/macro_particle_system.py:1038]: done

File: omnigibson/systems/macro_particle_system.py (hunk 12 above)
Comment (cremebrule) [reply omnigibson/systems/macro_particle_system.py:1084]: done

File: omnigibson/systems/macro_particle_system.py (hunk 13 above)
Comment (cremebrule) [reply omnigibson/systems/macro_particle_system.py:275]: removed

File: omnigibson/systems/macro_particle_system.py (hunk 15 above)
Comment (cremebrule) [reply omnigibson/systems/macro_particle_system.py:1311]: done

File: omnigibson/systems/macro_particle_system.py (hunk 18 above)
Comment (cremebrule) [reply omnigibson/systems/macro_particle_system.py:1325]: done

File: omnigibson/systems/macro_particle_system.py (hunk 17 above)
Comment (cremebrule) [reply omnigibson/systems/macro_particle_system.py]: done

File: omnigibson/systems/micro_particle_system.py (hunk 22 above)
Comment (cremebrule) [reply omnigibson/systems/micro_particle_system.py:1600]: added todo in asana, leave for future PR

File: omnigibson/object_states/contains.py [hunk 31]
```diff
@@ -56,23 +56,23 @@ def _get_value(self, system):
...
+                    unit_z[:, -1, :] = m.VISUAL_PARTICLE_OFFSET
+                    checked_positions = (T.quat2mat(quats) @ unit_z).reshape(-1, 3) + raw_positions
+            elif is_physical_particle_system(system_name=system.name):
+                raw_positions = system.get_particles_position_orientation()[0]
                 checked_positions = raw_positions
-        else:
-            raise ValueError(f"Invalid system {system} received for getting Covered state!"
-                             f"Currently, only VisualParticleSystems and PhysicalParticleSystems are supported.")
+            else:
+                raise ValueError(f"Invalid system {system} received for getting ContainedParticles state!"
```
Comment (ChengshuLi) [inline omnigibson/object_states/contains.py:74]: Nit: this should be `Contains` state?

File: omnigibson/systems/system_base.py (hunk 30 above)
[the remark this answers is not in the document]
Comment (ChengshuLi) [reply omnigibson/systems/system_base.py]: I think it should. Flat substances should have z_extent=0 so it doesn't matter how much you scale in the z-axis (i.e. we shouldn't treat z differently from x and y I think)

File: omnigibson/systems/system_base.py [hunk 32]
```diff
@@ -134,6 +432,569 @@ def __init__(self):
...
+        # doesn't already exist
+        assert group not in cls.groups, \
+            f"Cannot create new attachment group because group with name {group} already exists!"
+
+        # Create the group
+        cls._group_particles[group] = dict()
+        cls._group_objects[group] = obj
+
+        # Compute the group scale
+        cls._group_scales[group] = cls._compute_relative_group_scales(group=group)
```
Comment (ChengshuLi) [inline omnigibson/systems/system_base.py:586]: Should we only do this if cls.scale_relative_to_parent is True?

File: omnigibson/systems/system_base.py [hunk 33]
```diff
@@ -119,9 +220,206 @@ def get_active_systems(cls):
...
+        )
+
+    @classmethod
+    def _load_state(cls, state):
+        # Sanity check loading particles
+        assert cls.n_particles == state["n_particles"], f"Inconsistent number of particles found when loading " \
+                                                        f"particles state! Current number: {cls.n_particles}, " \
+                                                        f"loaded number: {state['n_particles']}"
+        # Load scale
+        cls.set_scale_limits(minimum=state["min_scale"], maximum=state["max_scale"])
```
Comment (ChengshuLi) [inline omnigibson/systems/system_base.py:390]: why do we need `set_scale_limits` at all? Can we just do 
```
cls.min_scale = state["min_scale"]
cls.max_scale = state["max_scale"]
```
Also, in load/dump_state, do we also need to handle `cls._group_scales`? Update: maybe it's handled by create/remove_attachment_group already.

File: omnigibson/systems/system_base.py [hunk 34]
```diff
@@ -119,9 +220,206 @@ def get_active_systems(cls):
         """
         return {system.name: system for system in SYSTEM_REGISTRY.objects if issubclass(system, cls)}
 
+    @classmethod
+    def set_scale_limits(cls, minimum=None, maximum=None):
+        """
+        Set the min and / or max scaling limits that will be uniformly sampled from when generating new particles.
+        Note: These scales dictate the scale set for particles in their parents' frame, NOT the global frame!
```
Comment (ChengshuLi) [inline omnigibson/systems/system_base.py:227]: Let's remove this function, and add some comments to the min_scale/max_scale variables. 

I think these are actually the desired scale for the particle templates in the global frame, not the parent's object frame.

File: omnigibson/systems/macro_particle_system.py [hunk 35]
```diff
@@ -73,6 +57,9 @@ def initialize(cls):
         og.sim.import_object(obj=particle_template, register=False)
         particle_template.kinematic_only = True
 
+        # Make sure the scale of the object is exactly 1.0
+        assert np.all(particle_template.scale == 1.0)
```
Comment (ChengshuLi) [inline omnigibson/systems/macro_particle_system.py:61]: sorry why is this required again?

File: omnigibson/systems/macro_particle_system.py [hunk 36]
```diff
@@ -155,95 +143,55 @@ def particle_name_prefix(cls):
...
-        return dict(
-            n_particles=cls.n_particles,
-            poses=[particle.get_local_pose() for particle in cls.particles.values()],
-            scales=[particle.scale for particle in cls.particles.values()],
-            template_pose=cls.particle_object.get_local_pose() if cls.particle_object is not None else None,
-            template_scale=cls.particle_object.scale if cls.particle_object is not None else None,
-        )
+        state = super()._dump_state()
+
+        state["scales"] = np.array([particle.scale for particle in cls.particles.values()])
```
Comment (ChengshuLi) [inline omnigibson/systems/macro_particle_system.py:154]: so do we allow MacroPhysicalParticleSystem to have non-uniform scale?

File: omnigibson/systems/macro_particle_system.py [hunk 37]
```diff
@@ -310,35 +230,64 @@ def add_particle(cls, prim_path, idn=None, scale=None, position=None, orientatio
...
+            scales=None,
+            **kwargs,
+    ):
+        # Grab pre-existing tfs
+        current_positions, current_orientations = cls.get_particles_position_orientation()
+
+        # Update the tensors
+        n_particles = len(positions)
+        orientations = R.random(num=n_particles).as_quat() if orientations is None else orientations
+        scales = cls.sample_scales(n=n_particles) if scales is None else scales
```
Comment (ChengshuLi) [inline omnigibson/systems/macro_particle_system.py:275]: sample_scales?

File: omnigibson/systems/macro_particle_system.py (hunk 4 above)
Comment (ChengshuLi) [reply omnigibson/systems/macro_particle_system.py:538]: Is this fixed?

File: omnigibson/systems/macro_particle_system.py [hunk 38]
```diff
@@ -952,10 +776,7 @@ def _sync_particle_groups(cls, group_objects, particle_idns, particle_attached_l
 
             for particle_idn, link_name in zip(info["particle_idns"], info["link_names"]):
                 # Create the necessary particles
-                particle = cls.add_particle(
-                    prim_path=f"{obj.prim_path}/{link_name}",
-                    idn=int(particle_idn),
-                )
+                particle = cls.add_particle(prim_path=f"{obj.prim_path}/{link_name}", scale=np.ones(3), idn=int(particle_idn))
```
Comment (ChengshuLi) [inline omnigibson/systems/macro_particle_system.py:779]: why scale=np.ones(3)?

File: omnigibson/systems/macro_particle_system.py [hunk 39]
```diff
@@ -1163,20 +1371,19 @@ def _deserialize(cls, state):
...
+        name=name,
+        # class_id=SemanticClass.DIRT,
+        visible=False,
+        fixed_base=False,
+        visual_only=True,
+        include_default_states=False,
+        category="raspberry",
+        model="spkers",
+    ),
+    scale=np.ones(3) * 5.0,
```
Comment (ChengshuLi) [inline omnigibson/systems/macro_particle_system.py:1388]: what is this scale doing again?

File: omnigibson/systems/micro_particle_system.py [hunk 40]
```diff
@@ -1610,12 +1522,12 @@ def cm_create_particle_template(cls):
         visual_only=True,
         include_default_states=False,
         abilities={},
+        scale=0.05,
```
Comment (ChengshuLi) [inline omnigibson/systems/micro_particle_system.py:1525]: What is this scale doing?

**Review by ChengshuLi — APPROVED** [review]

File: omnigibson/systems/macro_particle_system.py (hunk 4 above)
Comment (cremebrule) [reply omnigibson/systems/macro_particle_system.py:538]: fixed

File: omnigibson/systems/system_base.py (hunk 32 above)
Comment (cremebrule) [reply omnigibson/systems/system_base.py:586]: done

File: omnigibson/systems/system_base.py (hunk 33 above)
Comment (cremebrule) [reply omnigibson/systems/system_base.py:390]: done

File: omnigibson/systems/macro_particle_system.py (hunk 38 above)
Comment (cremebrule) [reply omnigibson/systems/macro_particle_system.py:779]: clarified

**Review by ChengshuLi — APPROVED** [review]
LGTM! Feel free to merge!
